cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/openshift/S2IOperation.java

```diff
@@ -1,19 +1,66 @@
 package io.strimzi.controller.cluster.operations.openshift;
 
+import io.strimzi.controller.cluster.OpenShiftUtils;
+import io.strimzi.controller.cluster.operations.Operation;
 import io.strimzi.controller.cluster.resources.Source2Image;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
 
 /**
  * Base Source2Image operation
  */
-public abstract class S2IOperation extends OpenShiftOperation {
+public abstract class S2IOperation implements Operation<OpenShiftUtils> {
+    private static final Logger log = LoggerFactory.getLogger(S2IOperation.class.getName());
+
     protected final Source2Image s2i;
+    private final String operationType;
 
     /**
      * Constructor
      *
      * @param s2i   Source2Image instance
      */
-    protected S2IOperation(Source2Image s2i) {
+    protected S2IOperation(String operationType, Source2Image s2i) {
+        this.operationType = operationType;
         this.s2i = s2i;
     }
+
+    @Override
+    public final void execute(Vertx vertx, OpenShiftUtils os, Handler<AsyncResult<Void>> handler) {
+        log.info("{} S2I {} in namespace {}", operationType, s2i.getName(), s2i.getNamespace());
+
+        try {
+            if (guard(os)) {
+                List<Future> futures = futures(os);
+
+                CompositeFuture.join(futures).setHandler(ar -> {
+                    if (ar.succeeded()) {
+                        log.info("S2I {} successfully updated in namespace {}", s2i.getName(), s2i.getNamespace());
+                        handler.handle(Future.succeededFuture());
+                    } else {
+                        log.error("S2I cluster {} failed to update in namespace {}", s2i.getName(), s2i.getNamespace());
+                        handler.handle(Future.failedFuture("Failed to update S2I"));
+                    }
+                });
+            } else {
+                log.info("No S2I {} differences found in namespace {}", s2i.getName(), s2i.getNamespace());
+                handler.handle(Future.succeededFuture());
+            }
+        }
+        catch (Exception e) {
+            log.error("S2I cluster {} failed to update in namespace {}", s2i.getName(), s2i.getNamespace(), e);
+            handler.handle(Future.failedFuture("Failed to update S2I"));
+        }
+    }
+
+    protected abstract boolean guard(OpenShiftUtils os);
```

Comment:
The `guard` method name seems to be strange. My first idea was that it is supposed to guard this against being called on Kubernetes. Maybe we should think about some better name.

Reply:
I agree the name is not great. It turns out we don't even need this method if we push the test down into the `UpdateS2IOperation`.